File: FrontEnd/src/utils/deviceStatus.ts
import { Device } from '../types/device';

/**
 * Determina el estado real de un dispositivo aplicando la lógica de prioridad:
 * 1. Si la nota contiene "Libre" → Libre (prioridad absoluta)
 * 2. Si no, usa iddevicestatus
 */
export const getDeviceRealStatus = (device: Device): 'libre' | 'operativo' | 'desconectado' | 'reducido' => {
    // PRIORIDAD 1: Nota con "Libre"
    if (device.strnote && device.strnote.toLowerCase().includes('libre')) {
        return 'libre';
    }

    // PRIORIDAD 2: iddevicestatus
    const statusId = device.iddevicestatus;

    if (statusId === 4) {
        return 'libre';
    } else if ([3, 8].includes(statusId)) {
        return 'desconectado';
    } else if ([2, 9].includes(statusId)) {
        return 'reducido';
    } else if ([1, 5, 6, 7].includes(statusId)) {
        return 'operativo';
    }

    // Fallback
    return 'desconectado';
};

/**
 * Obtiene la configuración visual para el estado de un dispositivo
 */
export const getStatusDisplay = (status: ReturnType<typeof getDeviceRealStatus>) => {
    switch (status) {
        case 'operativo':
            return { label: 'Operational', value: 'operational', color: 'green' };
        case 'desconectado':
            return { label: 'Disconnected', value: 'disconnected', color: 'red' };
        case 'reducido':
            return { label: 'Reduced', value: 'reduced', color: 'orange' };
        case 'libre':
            return { label: 'Libre', value: 'libre', color: 'purple' };
        default:
            return { label: 'Unknown', value: 'unknown', color: 'gray' };
    }
};

File: backend/src/utils/deviceStatus.py
"""
Utilidad para determinar el estado real de un dispositivo.
Replica la lógica del frontend (deviceStatus.ts) en Python.
"""

def get_device_real_status(device: dict) -> str:
    """
    Determina el estado real de un dispositivo aplicando la lógica de prioridad:
    1. Si la nota contiene "Libre" → Libre (prioridad absoluta)
    2. Si no, usa iddevicestatus
    
    Args:
        device: Diccionario con los datos del dispositivo
        
    Returns:
        str: 'libre', 'operativo', 'desconectado', o 'reducido'
    """
    # PRIORIDAD 1: Nota con "Libre"
    note = device.get('strnote', '')
    if note and 'libre' in note.lower():
        return 'libre'
    
    # PRIORIDAD 2: iddevicestatus
    status_id = device.get('iddevicestatus', 0)
    
    if status_id == 4:
        return 'libre'
    elif status_id in [3, 8]:
        return 'desconectado'
    elif status_id in [2, 9]:
        return 'reducido'
    elif status_id in [1, 5, 6, 7]:
        return 'operativo'
    
    # Fallback
    return 'desconectado'
